codegen/templates/http_client.tmpl

```diff
@@ -98,22 +99,79 @@ func (c *{{$clientName}}) {{title .Name}}(
 		respHeaders[k] = res.Header.Get(k)
 	}
 
-	res.CheckOKResponse({{.OKStatusCode.Code}})
+	res.CheckOKResponse([]int{
+		{{- range $index, $code := .ValidStatusCodes -}}
+		{{if $index}},{{end}}{{$code}}
+		{{- end -}}
+	})
 
-	{{if eq .ResponseType ""}}
+	{{if and (eq .ResponseType "") (eq (len .Exceptions) 0)}}
 	_, err = res.ReadAll()
 	if err != nil {
 		return respHeaders, err
 	}
 	return respHeaders, nil
-	{{else}}
+	{{else if eq .ResponseType ""}}
+	switch res.StatusCode {
+		{{range $idx, $exception := .Exceptions}}
+		case {{$exception.StatusCode.Code}}:
+			var exception {{$exception.Type}}
+			err = res.ReadAndUnmarshalBody(&exception)
+			if err != nil {
+				break
+			}
+			return respHeaders, &exception
+		{{end}}
+		default:
+			// TODO: log about unexpected body bytes?
+			_, err = res.ReadAll()
+	}
+
+	if err != nil {
+		return respHeaders, err
+	}
+	return respHeaders, nil
+	{{else if eq (len .Exceptions) 0}}
```

Comment:
Might be slightly more readable if this was the second case in your if-else-if-else chain. That way the two switch statements are together and the two "base" cases are together.